src/test/java/org/junit/platform/engine/test/event/ExecutionEventRecorder.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2015-2016 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.platform.engine.test.event;
+
+import static java.util.function.Predicate.isEqual;
+import static java.util.stream.Collectors.toList;
+import static org.junit.platform.commons.util.FunctionUtils.where;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.DYNAMIC_TEST_REGISTERED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.FINISHED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.REPORTING_ENTRY_PUBLISHED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.SKIPPED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.Type.STARTED;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byPayload;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byTestDescriptor;
+import static org.junit.platform.engine.test.event.ExecutionEvent.byType;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.junit.platform.engine.EngineDiscoveryRequest;
+import org.junit.platform.engine.EngineExecutionListener;
+import org.junit.platform.engine.ExecutionRequest;
+import org.junit.platform.engine.TestDescriptor;
+import org.junit.platform.engine.TestEngine;
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.TestExecutionResult.Status;
+import org.junit.platform.engine.UniqueId;
+import org.junit.platform.engine.reporting.ReportEntry;
+import org.junit.platform.engine.test.event.ExecutionEvent.Type;
+
+/**
+ * {@link EngineExecutionListener} that records all events and makes them available to tests.
+ *
+ * @since 1.0
+ * @see ExecutionEvent
+ */
+public class ExecutionEventRecorder implements EngineExecutionListener {
+
+	public static List<ExecutionEvent> execute(TestEngine testEngine, EngineDiscoveryRequest discoveryRequest) {
+		TestDescriptor engineTestDescriptor = testEngine.discover(discoveryRequest,
+			UniqueId.forEngine(testEngine.getId()));
+		ExecutionEventRecorder listener = new ExecutionEventRecorder();
+		testEngine.execute(
+			new ExecutionRequest(engineTestDescriptor, listener, discoveryRequest.getConfigurationParameters()));
+		return listener.getExecutionEvents();
+	}
+
+	public final List<ExecutionEvent> executionEvents = new CopyOnWriteArrayList<>();
```

Comment:
Have you considered using a functional persistent list here, like PCollections' `TreePVector` or Javaslang's `javaslang.collection.List`?